### DISCUSSION TITLE
Know which slices change when selecting multiple state slices

### DISCUSSION THREAD

**Discussion by ThomasRutzer** [pmndrs/zustand#851] [Q&A] [answered]
Hey, if I  am [selecting multiple state slices](https://github.com/pmndrs/zustand#selecting-multiple-state-slices) with an Array pick, is there any built-in solution to know which state slices actually change? Or do I need to store the previous value in my component and run a seperate comparison?

**Comment by dai-shi** [comment, marked as answer]
Yeah, you would need to store the previous value on your end. Alternatively, `.subscribe()` method may help. It depends on your requirements, but I can also think about storing the changes in the store.

**Reply by ThomasRutzer** [reply]
Hi @dai-shi, thanks for moving this from issues to discussions. And also thanks for the super quick answer. 

Unfortunatly, `subscribe` is not solving the case for me, I need the re-render.  It's a very special requirement on my end: 

I generate a list of dependencies for an `useEffect` like this:

```
  const sendDeps = useStore(state => syncPaths.map(key => get(state, key)), shallow) // syncPaths is a list with any store slice 

  useEffect(() => {
    const sendState = syncPaths.reduce(
      (prev, curr) => ({
        ...prev,
        [curr]: get(useStore.getState(), curr),
      }),
      {},
    )

   doSendData(sendState)
  }, sendDeps)
```

`doSendData` actually sends the changes over network, but for performance reasons, I'd like to only send the diff. So I was hoping there is an easy solution, since `shallow` at least knows which keys changed. But it seems that I need to repeat this once again in the effect. 

Thank you so far!

**Reply by dai-shi** [reply]
Assuming you don't use sendDeps to render anything, I'd do something like this:
```js
  useEffect(() => useStore.subscribe((state, prev) => {
    ...
  }, [syncPaths])
```

**Reply by ThomasRutzer** [reply]
Ok this was a good pointer. I just discovered that `subscribe` with a selector can also be used to handle a state pick. And it also exposes previous state, so together with a custom comparison on my end, I think this works.

Thank you a lot, helping me out here and with `Zustand` in general. It is just awesome!
